rtt_ros2_params/src/rtt_ros2_params.cpp

```diff
@@ -0,0 +1,296 @@
+      prop = getProperty(property_name);
+    } else {
+      orphan_params_[property_name] = paramvalue;
+      prop = getProperty(property_name);
+    }
+    return true;
+  } else {
+    RTT::Property<rclcpp::ParameterValue> prop_paramvalue = prop;
+
+    // The property exists on the Owner's interface
+    RTT::log(RTT::Debug) <<
+      "The property " << property_name << " existed" <<
+      RTT::endlog();
+    if (!prop_paramvalue.ready()) {
+      RTT::log(RTT::Debug) <<
+        "But " << property_name <<
+        " cannot be casted into rclcpp::ParamValue" << RTT::endlog();
+
+      const auto property_ds = prop->getDataSource();
+      RTT::internal::ReferenceDataSource<rclcpp::ParameterValue> param_rds(paramvalue);
+
+      try {
+        if (property_ds->update(&param_rds)) {
+          // conversion successful
+          RTT::log(RTT::Debug) <<
+            "Property " <<
+            getOwner()->provides()->getName() << "." << property_name <<
+            " loaded successfully from ROS2 parameter " << param_name << RTT::endlog();
+          return true;
+        } else {
+          // conversion failed
+          RTT::log(RTT::Warning) <<
+            "Property " <<
+            getOwner()->provides()->getName() << "." << property_name <<
+            " failed to convert from ROS2 parameter " << param_name << RTT::endlog();
+          return false;
+        }
+      } catch (std::exception e) {
+        RTT::log(RTT::Error) <<
+          "The property " << getOwner()->provides()->getName() << "." <<
+          property_name << " could not be converted from ROS2 parameter " << param_name <<
+          ". Exception caught: " << e.what() << RTT::endlog();
+        return false;
+      }
+      return false;
+    } else {
+      // This is the rare case in which the Property was already of type rclcpp::ParameterValue
+      prop_paramvalue.set(paramvalue);
+    }
+    return true;
+  }
+  return false;
+}
+
+bool Params::storeProperty(const std::string property_name, const std::string param_name)
+{
+  RTT::Logger::In in(getName());
+
+  const RTT::base::PropertyBase * prop = getOwner()->properties()->getProperty(property_name);
+  if (nullptr != prop) {
+    RTT::log(RTT::Debug) <<
+      getOwner()->provides()->getName() <<
+      " has a property " << property_name << RTT::endlog();
+  } else {
+    RTT::log(RTT::Error) <<
+      getOwner()->provides()->getName() <<
+      " has NO property " << property_name << RTT::endlog();
+    for (const auto property : getOwner()->properties()->getProperties() ) {
+      RTT::log(RTT::Debug) <<
+        "Available property: " << property->getName() <<
+        RTT::endlog();
+    }
+    return false;
+  }
+
+  RTT::base::DataSourceBase::shared_ptr property_bds = prop->getDataSource();
+  RTT::internal::ValueDataSource<rclcpp::ParameterValue> param_vds;
+
+  // TODO(spdintermodalics): check whether the conversion was successful!
+  param_vds.update(property_bds.get());
```

Comment:
I added this comment in https://github.com/orocos/rtt_ros2_integration/pull/7/commits/60239c36a5470818d6ce9b8da78043f5708367b2. Also for `storeProperty()` the `update()` might fail if the property has an incompatible type that cannot be converted to `ParameterValue`. We must not set the parameter with a default-constructed `rclcpp::ParameterValue` (the default value of `ValueDataSource<rclcpp::ParameterValue>`. Or was it like that on purpose, because the default value of `ParameterValue` is "not set"? I think `set_parameter()` even deletes the ROS parameter in this case.

Reply:
I made it this way because the parameter type is assigned depending on the contents when you construct it. At least, when I was checking the type after calling `storeProperty()` when no parameter was existing, the parameter type assigned was correct.

Reply:
But `update()` will fail if the property has a type that cannot be converted to `rclcpp::ParameterValue` by the typekit. So basically for all types but integer and floating point types, `bool` and `string`. An RTT property could be of type `KDL::Frame`, for example.